classes/notification/NotificationDAO.inc.php

```diff
@@ -234,25 +234,20 @@ public function build($contextId, $level, $type, $assocType, $assocId, $userId =
      * @param int $notificationId
      * @param int $userId
      *
-     * @return bool
      */
     public function deleteById($notificationId, $userId = null)
     {
-        $params = [(int) $notificationId];
-        if (isset($userId)) {
-            $params[] = (int) $userId;
-        }
-        if ($this->update(
-            'DELETE FROM notifications WHERE notification_id = ?' . (isset($userId) ? ' AND user_id = ?' : ''),
-            $params
-        )) {
-            // If a notification was deleted (possibly validating
-            // $userId in the process) delete associated settings.
-            $notificationSettingsDao = DAORegistry::getDAO('NotificationSettingsDAO'); /** @var NotificationSettingsDAO $notificationSettingsDaoDao */
-            $notificationSettingsDao->deleteSettingsByNotificationId($notificationId);
-            return true;
+        $notificationSettingsDao = DAORegistry::getDAO('NotificationSettingsDAO'); /** @var NotificationSettingsDAO $notificationSettingsDaoDao */
+        $notificationSettingsDao->deleteSettingsByNotificationId($notificationId);
+
+        $query = DB::table('notifications')
+            ->where('notification_id', '=', $notificationId);
+        
+        if ($userId) {
+            $query->where('user_id', '=', $userId);
         }
-        return false;
+
+        $query->delete();
```

Comment:
If the `$userId` doesn't match the notification ID, this will cause the settings to be deleted but the notification to remain.

Reply:
This is a general risk with the way you've implemented a few delete functions here -- please review. You might need to...
- Fetch the object by whatever IDs are provided
- If it exists...
  - Delete any dependent records that
    - are not covered by foreign keys with `ON DELETE CASCADE` SQL clauses, *or*
    - have dependent elements that need manual consideration, like files in the filesystem needing removal
  - Lastly, delete the object itself.

As you can see, having foreign keys declared with `ON DELETE` clauses is helpful in many/most cases! Consider adding these rather than manually implementing deletion of dependent elements.

Reply:
I see - it makes sense. I tried to make the appropriate changes. These are relatively new for me, so could you have another round of review if you like?